coqpyt/coq/base_file.py

```diff
@@ -260,93 +256,173 @@ def _make_change(self, change_function, *args):
-        ):
-            raise NotImplementedError(
-                "Deleting steps outside of a proof is not implemented yet"
-            )
+    def __delete_step_text(self, step_index: int):
+        with open(self.__path, "r") as f:
+            lines = f.readlines()
+
+        deleted_step = self.steps[step_index]
+        step = deleted_step
+        if step_index != 0:
+            prev_step_end = self.steps[step_index - 1].ast.range.end
+        else:
+            prev_step_end = Position(0, 0)
+
+        start_line = lines[prev_step_end.line]
+        end_line = lines[step.ast.range.end.line]
+
+        start_line = start_line[: prev_step_end.character]
+        end_line = end_line[step.ast.range.end.character :]
+
+        if prev_step_end.line == step.ast.range.end.line:
+            lines[prev_step_end.line] = start_line + end_line
+        else:
+            lines[prev_step_end.line] = start_line
+            lines[step.ast.range.end.line] = end_line
 
+        # Delete lines between first and last line
+        for i in range(prev_step_end.line + 1, step.ast.range.end.line):
+            del lines[i]
+        text = "".join(lines)
+
+        with open(self.__path, "w") as f:
+            f.write(text)
+
+    def __add_step_text(self, previous_step_index: int, step_text: str):
         with open(self.__path, "r") as f:
             lines = f.readlines()
 
-        with open(self.path, "w") as f:
-            step = self.steps[step_index]
-            prev_step = self.steps[step_index - 1]
-            start_line = lines[prev_step.ast.range.end.line]
-            end_line = lines[step.ast.range.end.line]
+        previous_step = self.steps[previous_step_index]
+        end_line = lines[previous_step.ast.range.end.line]
+        end_line = (
+            end_line[: previous_step.ast.range.end.character]
+            + step_text
+            + end_line[previous_step.ast.range.end.character :]
+        )
+        lines[previous_step.ast.range.end.line] = end_line
 
-            start_line = start_line[: prev_step.ast.range.end.character]
-            end_line = end_line[step.ast.range.end.character :]
+        text = "".join(lines)
 
-            if prev_step.ast.range.end.line == step.ast.range.end.line:
-                lines[prev_step.ast.range.end.line] = start_line + end_line
-            else:
-                lines[prev_step.ast.range.end.line] = start_line
-                lines[step.ast.range.end.line] = end_line
-            text = "".join(lines)
+        with open(self.__path, "w") as f:
             f.write(text)
 
+    def __delete_update_ast(self, step_index: int):
+        deleted_step = self.steps[step_index]
+        for step in self.steps[step_index + 1 :]:
+            remove_chars = deleted_step.ast.range.end.character
+            if deleted_step.ast.range.end.line == deleted_step.ast.range.start.line:
+                remove_chars -= deleted_step.ast.range.start.character - 1
```

Comment:
Yes, I think you are partially right, it's not because of the `\n` tho, it's because of a white space.

This is tested on `TestProofSimpleFileChanges`. Namely, when removing the reflexivity:
```
Example test1: 1 + 1 = 2. reflexivity. Qed.
```
The reflexivity step starts on character 26 and ends on 38, and that does not consider the white space that I think that we also remove. And that's why I added the -1. Because that way, the remove_chars will be 13 instead of 12.

We have to test it, but I think your change is correct. However, when doing that change we have to check if the step_index is 0 or not, and if it is, use (0,0) as a replacement of `prev_step.ast.range.end`.